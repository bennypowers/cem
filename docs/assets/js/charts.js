import { setBasePath } from 'https://cdn.jsdelivr.net/npm/@shoelace-style/shoelace@2.16.0/dist/utilities/base-path.js';
setBasePath('https://cdn.jsdelivr.net/npm/@shoelace-style/shoelace@2.16.0/dist/');

// Theme switching is now handled by the color-mode-toggle web component

// JSON viewer lazy loading setup
document.addEventListener('DOMContentLoaded', () => {
  const disclosures = document.querySelectorAll('.json-disclosure');
  
  disclosures.forEach(disclosure => {
    const details = disclosure;
    const spinner = details.querySelector('sl-spinner');
    const viewer = details.querySelector('json-viewer');
    const jsonUrl = details.dataset.jsonUrl;
    let loaded = false;
    
    details.addEventListener('sl-show', async () => {
      if (loaded) return;
      
      try {
        const response = await fetch(jsonUrl);
        const jsonData = await response.json();
        
        viewer.data = jsonData;
        viewer.style.display = 'block';
        spinner.style.display = 'none';
        loaded = true;
      } catch (error) {
        console.error('Failed to load JSON:', error);
        spinner.innerHTML = 'Failed to load JSON data';
      }
    });
  });
});

export class BarChart extends HTMLElement {
  static is = 'bar-chart';
  static template = document.createElement('template');
  static { this.template.innerHTML = `<slot></slot>`; }
  static { customElements.define(this.is, this); }
  constructor() {
    super()
    this.attachShadow({ mode: 'open' }).append(BarChart.template.content.cloneNode(true));
  }
}

export class LineChart extends HTMLElement {
  static is = 'line-chart';
  static template = document.createElement('template');
  static {
    this.template.innerHTML = `
      <style>
        :host {
          /* Chart variables */
          --chart-axis-stroke: var(--sl-panel-border-color, #888);
          --chart-label-color: var(--sl-color-neutral-900, currentColor);
          --chart-line-stroke: var(--sl-color-primary-600, #4e79a7);
          --chart-point-fill: var(--sl-tooltip-background-color, #b3c9e5);
          --chart-point-stroke: var(--sl-tooltip-border-color, #4e79a7);
          --chart-point-highlight: var(--sl-color-primary-400, #7ea7d8);
          --chart-point-highlight-stroke: var(--sl-color-primary-700, #22527a);
          /* Tooltip variables */
          --tooltip-background: var(--sl-tooltip-background-color, #222);
          --tooltip-color: var(--sl-tooltip-color, #fff);
          --tooltip-border-radius: var(--sl-tooltip-border-radius, 4px);
          --tooltip-border-color: var(--sl-tooltip-border-color, #4e79a7);
          --tooltip-shadow: var(--sl-tooltip-shadow, 0 2px 8px 0 #2222);
          --tooltip-padding: var(--sl-spacing-x-small, 4px);
        }
        .chart-tooltip {
          position: fixed;
          background: var(--tooltip-background);
          color: var(--tooltip-color);
          padding: var(--tooltip-padding);
          border-radius: var(--tooltip-border-radius);
          border: 1.5px solid var(--tooltip-border-color);
          box-shadow: var(--tooltip-shadow);
          pointer-events: none;
          font-size: 13px;
          z-index: 10000;
          display: none;
          transition: opacity 0.1s;
        }
        .run-point {
          transition: r 0.1s, fill 0.1s, stroke 0.1s;
          cursor: pointer;
        }
        .run-point[data-hovered] {
          fill: var(--chart-point-highlight, #7ea7d8);
          stroke: var(--chart-point-highlight-stroke, #22527a);
          r: 6;
        }
      </style>
    `;
    customElements.define(this.is, this);
  }

  #tooltip;
  #points;
  #svg;
  #hoveredIndex = null;

  constructor() {
    super();
    this.attachShadow({ mode: 'open' }).append(LineChart.template.content.cloneNode(true));
    this.#tooltip = null;
    this.#points = [];
    this.#svg = null;
  }

  connectedCallback() {
    // copy svg into shadowroot
    this.#svg = this.querySelector('svg');
    this.shadowRoot.append(this.#svg);
    if (!this.#svg) return;
    this.#points = Array.from(this.#svg.querySelectorAll('circle.run-point')).map((pt, i) => ({
      el: pt,
      x: parseFloat(pt.getAttribute('cx')),
      y: parseFloat(pt.getAttribute('cy')),
      run: +pt.dataset.run,
      time: pt.dataset.time,
      i
    }));
    this.#svg.addEventListener('pointermove', this.#onPointerMove.bind(this));
    this.#svg.addEventListener('pointerleave', this.#onPointerOut.bind(this));
  }

  #onPointerMove(e) {
    if (!this.#svg || !this.#points.length) return;

    // Convert mouse to SVG coordinates
    const pt = this.#svg.createSVGPoint();
    pt.x = e.clientX;
    pt.y = e.clientY;
    const svgPt = pt.matrixTransform(this.#svg.getScreenCTM().inverse());
    const mouseX = svgPt.x;
    const mouseY = svgPt.y;

    // Find closest point (2D)
    let closest = this.#points[0];
    let minDist = Math.hypot(mouseX - closest.x, mouseY - closest.y);
    for (const pt of this.#points) {
      const dist = Math.hypot(mouseX - pt.x, mouseY - pt.y);
      if (dist < minDist) {
        closest = pt;
        minDist = dist;
      }
    }

    // Highlight and show tooltip as before...
    if (this.#hoveredIndex !== null && this.#points[this.#hoveredIndex]) {
      this.#points[this.#hoveredIndex].el.removeAttribute('data-hovered');
    }
    this.#hoveredIndex = closest.i;
    closest.el.setAttribute('data-hovered', '');

    if (!this.#tooltip) {
      this.#tooltip = document.createElement('div');
      this.#tooltip.className = 'chart-tooltip';
      this.shadowRoot.appendChild(this.#tooltip);
    }
    this.#tooltip.textContent = `Run #${closest.run}: ${closest.time}s`;
    // Position tooltip (you may want to clamp to stay inside viewport)
    const svgRect = this.#svg.getBoundingClientRect();
    this.#tooltip.style.left = `${svgRect.left + closest.x + 12}px`;
    this.#tooltip.style.top = `${svgRect.top + closest.y - 18}px`;
    this.#tooltip.style.display = 'block';
    this.#tooltip.style.opacity = '1';
  }
    #onPointerOut() {
    if (this.#tooltip) {
      this.#tooltip.style.display = 'none';
      this.#tooltip.style.opacity = '0';
    }
    if (this.#hoveredIndex !== null && this.#points[this.#hoveredIndex]) {
      this.#points[this.#hoveredIndex].el.removeAttribute('data-hovered');
    }
    this.#hoveredIndex = null;
  }
}