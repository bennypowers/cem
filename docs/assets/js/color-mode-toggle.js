const modes = ['light', 'auto', 'dark'];

export class ColorModeToggle extends HTMLElement {
  static is = 'color-mode-toggle';
  static { customElements.define(this.is, this); }

  #currentMode = 'auto';
  #storageKey = 'theme';

  connectedCallback() {
    this.#initializeMode();
    this.#setupEventListeners();
  }

  #initializeMode() {
    // Get stored preference or default to auto
    const stored = localStorage.getItem(this.#storageKey);
    this.#currentMode = modes.includes(stored) ? stored : 'auto';
    this.#applyMode(this.#currentMode);
    this.#updateButtons();
  }

  #setupEventListeners() {
    this.shadowRoot.addEventListener('change', this.#onChange.bind(this));
    
    // Listen for system color scheme changes
    const mediaQuery = window.matchMedia('(prefers-color-scheme: dark)');
    mediaQuery.addEventListener('change', () => {
      if (this.#currentMode === 'auto') {
        this.#applyMode('auto');
      }
    });
  }

  #onChange(event) {
    if (event.target.type === 'radio' && event.target.name === 'color-mode') {
      const mode = event.target.value;
      if (mode && mode !== this.#currentMode) {
        this.#setMode(mode);
      }
    }
  }

  #setMode(mode) {
    this.#currentMode = mode;
    localStorage.setItem(this.#storageKey, mode);
    this.#applyMode(mode);
    this.#updateButtons();
    
    // Dispatch custom event for other components
    this.dispatchEvent(new CustomEvent('color-mode-change', {
      detail: { mode },
      bubbles: true
    }));
  }

  #applyMode(mode) {
    const html = document.documentElement;
    
    // Remove existing mode classes/attributes
    html.removeAttribute('data-mode');
    html.classList.remove('light', 'dark');
    
    switch (mode) {
      case 'light':
        html.setAttribute('data-mode', 'light');
        html.classList.add('light');
        break;
      case 'dark':
        html.setAttribute('data-mode', 'dark');
        html.classList.add('dark');
        break;
      case 'auto':
        html.setAttribute('data-mode', 'auto');
        // Let CSS handle prefers-color-scheme
        break;
    }

    // Update pictures for current mode
    this.#updatePictures(this.#getEffectiveMode(mode));
  }

  #getEffectiveMode(mode) {
    if (mode === 'auto') {
      return window.matchMedia('(prefers-color-scheme: dark)').matches ? 'dark' : 'light';
    }
    return mode;
  }

  #updatePictures(effectiveMode) {
    document.querySelectorAll('picture[data-lit][data-dark]').forEach(picture => {
      const source = picture.firstElementChild;
      if (source && source.tagName === 'IMG') {
        const lightSrc = picture.dataset.lit;
        const darkSrc = picture.dataset.dark;
        source.src = effectiveMode === 'dark' ? darkSrc : lightSrc;
      }
    });
  }

  #updateButtons() {
    const radios = this.shadowRoot.querySelectorAll('input[type="radio"]');
    radios.forEach(radio => {
      radio.checked = radio.value === this.#currentMode;
    });
  }

}
