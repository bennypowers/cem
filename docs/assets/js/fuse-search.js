import Fuse from 'fuse.js';

export class FuseSearch extends HTMLElement {
  static #searchKeys = ['body', 'title', 'link', 'section', 'id'];

  static #minQueryLength = 2;

  static is = 'fuse-search'
  static { customElements.define(this.is, this); }

  #fuse = null;
  #searchInput = null;
  #resultsContainer = null;
  #form = null;
  #searchData = [];
  #selectedIndex = -1;
  #debounceTimer = null;

  connectedCallback() {
    this.#initializeElements();
    this.#setupEventListeners();
    this.#loadSearchData();
  }

  #initializeElements() {
    const shadowRoot = this.shadowRoot;
    this.#searchInput = shadowRoot.querySelector('input[type="search"]');
    this.#resultsContainer = shadowRoot.querySelector('[role="listbox"]');
    this.#form = shadowRoot.querySelector('form');
  }

  #setupEventListeners() {
    // Prevent form submission and use Fuse search instead
    this.#form.addEventListener('submit', this.#handleFormSubmit.bind(this));

    // Handle search input
    this.#searchInput.addEventListener('input', this.#handleSearchInput.bind(this));

    // Handle keyboard navigation
    this.#searchInput.addEventListener('keydown', this.#handleKeyDown.bind(this));

    // Clear results when clicking outside
    document.addEventListener('click', this.#handleDocumentClick.bind(this));

    // Clear results on escape
    document.addEventListener('keydown', (event) => {
      if (event.code === 'Escape') {
        this.#clearResults();
        this.#searchInput.blur();
      }
    });
  }

  #handleFormSubmit(event) {
    if (this.#fuse && this.#searchInput.value.trim()) {
      // Prevent default form submission if Fuse is available
      event.preventDefault();
      return;
    }
    // Allow default form submission to DuckDuckGo if Fuse not available
  }

  #handleSearchInput(event) {
    const query = event.target.value.trim();

    // Clear existing debounce timer
    if (this.#debounceTimer) {
      clearTimeout(this.#debounceTimer);
    }

    // Debounce search to avoid excessive API calls
    this.#debounceTimer = setTimeout(() => {
      this.#performSearch(query);
    }, 150);
  }

  #handleKeyDown(event) {
    const results = this.#resultsContainer.querySelectorAll('a');

    switch (event.code) {
      case 'ArrowDown':
        event.preventDefault();
        this.#selectedIndex = Math.min(this.#selectedIndex + 1, results.length - 1);
        this.#updateSelection(results);
        break;

      case 'ArrowUp':
        event.preventDefault();
        this.#selectedIndex = Math.max(this.#selectedIndex - 1, -1);
        this.#updateSelection(results);
        break;

      case 'Enter':
        event.preventDefault();
        if (this.#selectedIndex >= 0 && results[this.#selectedIndex]) {
          window.location.href = results[this.#selectedIndex].href;
        }
        break;

      case 'Escape':
        this.#clearResults();
        this.#searchInput.blur();
        break;
    }
  }

  #handleDocumentClick(event) {
    if (!this.contains(event.target)) {
      this.#clearResults();
    }
  }

  #updateSelection(results) {
    results.forEach((result, index) => {
      result.classList.toggle('active', index === this.#selectedIndex);
    });
  }

  async #loadSearchData() {
    try {
      const pageLanguage = document.documentElement.lang || 'en';
      const searchIndex = pageLanguage === 'en' ? 'index.json' : `${pageLanguage}/index.json`;
      const baseUrl = this.getAttribute('base-url') || window.location.origin;

      const response = await fetch(new URL(searchIndex, baseUrl).href);

      if (!response.ok) {
        throw new Error(`Failed to load search index: ${response.status}`);
      }

      this.#searchData = await response.json();

      // Initialize Fuse with the loaded data
      this.#fuse = new Fuse(this.#searchData, {
        ignoreLocation: true,
        findAllMatches: true,
        includeScore: true,
        shouldSort: true,
        keys: FuseSearch.#searchKeys,
        threshold: 0.1
      });

    } catch (error) {
      console.warn('Search index could not be loaded, falling back to external search:', error);
      // Fuse search will remain unavailable, form will submit to DuckDuckGo
    }
  }

  #performSearch(query) {
    if (!query || query.length < FuseSearch.#minQueryLength) {
      this.#clearResults();
      return;
    }

    if (!this.#fuse) {
      // Show loading state or fallback message
      this.#showMessage('Search index loading...');
      return;
    }

    const results = this.#fuse.search(query);
    this.#displayResults(results, query);
  }

  #displayResults(results, query) {
    this.#clearResults();
    this.#selectedIndex = -1;

    if (results.length === 0) {
      this.#showMessage('No results found');
      return;
    }

    // Limit results for performance
    const limitedResults = results.slice(0, 8);

    limitedResults.forEach(({ item, score }) => {
      const resultElement = this.#createResultElement(item, query, score);
      this.#resultsContainer.appendChild(resultElement);
    });

    this.#resultsContainer.classList.add('active');
  }

  #createResultElement(item, query) {
    const link = document.createElement('a');
    link.href = `${item.link}?query=${encodeURIComponent(query)}`;

    const title = document.createElement('h3');
    title.textContent = item.title;
    link.appendChild(title);

    // Add description if available
    if (item.body) {
      const description = document.createElement('p');
      const queryIndex = item.body.toLowerCase().indexOf(query.toLowerCase());
      const start = Math.max(0, queryIndex - 50);
      const end = Math.min(item.body.length, queryIndex + 150);
      const excerpt = item.body.substring(start, end);
      description.textContent = start > 0 ? `...${excerpt}` : excerpt;
      link.appendChild(description);
    }

    return link;
  }

  #showMessage(message) {
    this.#clearResults();

    const messageElement = document.createElement('div');
    messageElement.setAttribute('role', 'status');
    messageElement.textContent = message;

    this.#resultsContainer.appendChild(messageElement);
    this.#resultsContainer.classList.add('active');
  }

  #clearResults() {
    this.#resultsContainer.innerHTML = '';
    this.#resultsContainer.classList.remove('active');
    this.#selectedIndex = -1;
  }
}
