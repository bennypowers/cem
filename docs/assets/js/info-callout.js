class InfoCallout extends HTMLElement {
  constructor() {
    super();
    this.setupTitle();
    this.setupCollapsible();
  }

  setupTitle() {
    const title = this.getAttribute('title');
    if (title) {
      const titleEl = this.shadowRoot.querySelector('.title');
      titleEl.textContent = title;
    }
  }

  setupCollapsible() {
    if (this.hasAttribute('collapsible')) {
      const header = this.shadowRoot.querySelector('.header');
      header.addEventListener('click', () => {
        this.toggleAttribute('expanded');
      });
    } else {
      this.setAttribute('expanded', '');
    }
  }
}

if (!customElements.get('info-callout')) {
  customElements.define('info-callout', InfoCallout);
}