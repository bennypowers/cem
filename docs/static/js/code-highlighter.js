import {
  translations
} from './config.js';

// SVG icon loader
function loadSvg(icon, parent) {
  parent.innerHTML = `<svg class="icon_${icon}"><use xlink:href="#${icon}"></use></svg>`;
}

export class CodeHighlighter extends HTMLElement {
  static is = 'code-highlighter';
  static { customElements.define(this.is, this); }

  // Get CSS custom properties for better maintainability
  get maxHeight() {
    return getComputedStyle(document.documentElement).getPropertyValue('--code-max-height').trim() || 'initial';
  }

  get panelTimeout() {
    return parseInt(getComputedStyle(document.documentElement).getPropertyValue('--code-panel-timeout')) || 2250;
  }

  get animationDelay() {
    return parseInt(getComputedStyle(document.documentElement).getPropertyValue('--code-panel-animation-delay')) || 50;
  }

  connectedCallback() {
    this.#initializeCodeBlock();
    this.#setupEventListeners();
  }

  #initializeCodeBlock() {
    const codeElement = this.querySelector('code');
    if (!codeElement) return;

    this.#addLines(codeElement);
    this.#addActionPanel();
    this.#addLanguageLabel(codeElement);
  }

  #addLines(block) {
    let text = block.textContent;
    const snippet_fragment = [];
    if (text.includes('\n') && block.closest('pre') && !block.children.length) {
      text = text.split('\n');
      text.forEach((text_node, index) => {
        if(text_node.trim().length) {
          const new_node = `
          <span class="line line-flex">
            <span class="ln">${index + 1}</span>
            <span class="cl">${text_node.trim()}</span>
          </span>`.trim();
          snippet_fragment.push(new_node);
          block.closest('pre').className = 'chroma';
          block.classList.add('language-unknown');
          block.dataset.lang = translations.not_set;
        }
      });
      block.innerHTML = snippet_fragment.join('').trim(' ');
    }
  }


  #addActionPanel() {
    const actions = [
      { icon: 'copy', id: 'copy', title: translations.copy_text, show: true },
      { icon: 'order', id: 'lines', title: translations.toggle_line_numbers_text, show: true },
      { icon: 'carly', id: 'wrap', title: translations.toggle_line_wrap_text, show: false }
    ];

    const panel = document.createElement('div');
    panel.className = 'panel_box';

    actions.forEach(action => {
      const btn = document.createElement('a');
      btn.href = '#';
      btn.title = action.title;
      btn.className = `icon panel_icon panel_${action.id}`;
      if (!action.show) {
        btn.classList.add('panel_hide');
      }
      loadSvg(action.icon, btn);
      panel.appendChild(btn);
    });

    this.appendChild(panel);
  }


  #addLanguageLabel(block) {
    let label = block.dataset.lang;
    const shellBased = ['sh', 'shell', 'zsh', 'bash'];
    if (shellBased.includes(label)) {
      // Add shell class to the code block for CSS styling
      block.classList.add('language-shell');
      // Clean up shell prompts - remove $ from all lines since CSS will add it to first line
      const lines = block.querySelectorAll('.line .cl');
      Array.from(lines).forEach(line => {
        let content = line.textContent;
        // Remove leading $ and optional space
        if (content.startsWith('$ ')) {
          line.textContent = content.substring(2);
        } else if (content.startsWith('$')) {
          line.textContent = content.substring(1);
        }
      });
    }

    label = label === 'sh' ? 'shell' : label;
    if (label !== "fallback") {
      const labelEl = document.createElement('div');
      labelEl.textContent = label;
      labelEl.classList.add('lang');
      this.appendChild(labelEl);
    }
  }

  #setupEventListeners() {
    this.addEventListener('click', this.#handleClick.bind(this));
  }

  #handleClick(event) {
    const target = event.target;
    const action = this.#getActionType(target);
    if (!action) return;
    event.preventDefault();
    this.#showActiveState(target);
    const codeElement = this.querySelector('code');
    if (!codeElement) return;
    switch (action) {
      case 'copy':
        this.#copyCode(codeElement);
        break;
      case 'wrap':
        this.#toggleLineWrap(codeElement);
        break;
      case 'lines':
        this.#toggleLineNumbers(codeElement);
        break;
    }
  }

  #getActionType(target) {
    const classes = ['panel_copy', 'panel_wrap', 'panel_lines'];
    for (const cls of classes) {
      if (target.matches(`.${cls}`) || target.closest(`.${cls}`)) {
        return cls.replace('panel_', '');
      }
    }
    return null;
  }

  #showActiveState(target) {
    const element = target.matches('.icon') ? target : target.closest('.icon');
    if (!element) return;
    element.classList.remove('active');
    setTimeout(() => {
      element.classList.toggle('active');
    }, this.animationDelay);
  }

  #copyCode(codeElement) {
    const copyBtn = this.querySelector('.panel_copy');
    if (!copyBtn) return;

    const originalTitle = copyBtn.title;
    loadSvg('check', copyBtn);
    copyBtn.title = translations.copied_text;

    // Clone and clean up code for copying
    const clonedCode = codeElement.cloneNode(true);
    const lineNumbers = clonedCode.querySelectorAll('.ln');
    lineNumbers.forEach(ln => ln.remove());

    // Remove leading '$' from shell snippets
    const spans = clonedCode.querySelectorAll('span');
    spans.forEach(span => {
      const text = span.textContent.trim(' ');
      if (text.indexOf('$') === 0) {
        span.textContent = span.textContent.replace('$ ', '');
      }
    });

    const snippet = clonedCode.textContent.trim(' ');
    navigator.clipboard.writeText(snippet);

    setTimeout(() => {
      copyBtn.title = originalTitle;
      loadSvg('copy', copyBtn);
    }, this.panelTimeout);
  }

  #toggleLineWrap(codeElement) {
    codeElement.classList.toggle('pre_wrap');
  }

  #toggleLineNumbers(codeElement) {
    const lineNumbers = codeElement.querySelectorAll('.ln');
    lineNumbers.forEach(ln => ln.parentElement?.classList.toggle('pre_nolines'));
  }
}

