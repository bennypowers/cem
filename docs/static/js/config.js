// Configuration module with Hugo templating
// This module contains all values that need Hugo template processing

// Site configuration
export const site_title = `cem`;
export let root_url = '/cem/';
root_url = root_url.startsWith('http') ? root_url : window.location.origin;

// Hugo configuration values
export const code_block_config = {"maximum": 10, "show": true};
export const iconsPath = `icons/`;

// Derived configuration
export const max_lines = code_block_config.maximum;
export const show_lines = code_block_config.show;

// Theme configuration with Hugo values
export const storageKey = `${site_title}-color-mode`;
export const mermaidThemeKey = `${site_title}-mermaid`;
