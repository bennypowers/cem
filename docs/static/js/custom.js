// Add custom site-specific JavaScript here
// This module will be bundled with the main application

export function customInit() {
  // Add any custom initialization code here
  console.log('Custom JS module loaded');
}