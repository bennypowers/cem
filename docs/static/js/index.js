import { translations } from './config.js';

// Navigation toggle
function initializeNavigation() {
  const navToggle = document.getElementById('nav-toggle');
  navToggle?.addEventListener('click', function() {
    this.parentElement?.querySelector('nav')?.classList.toggle('show');
  });
}

function copyFeedback(parent) {
  const copy_txt = document.createElement('div');
  const yanked = 'link_yanked';
  copy_txt.classList.add(yanked);
  copy_txt.innerText = translations.copied_text;
  if(!parent.querySelector(`.${yanked}`)) {
    const icon = parent.getElementsByTagName('svg')?.[0];
    if (icon) {
      const original_src = icon.src;
      icon.src = '/icons/check.svg';
      parent.appendChild(copy_txt);
      setTimeout(function() {
        parent.removeChild(copy_txt)
        icon.src = original_src;
      }, 2250);
    }
  }
}

function copyHeadingLink() {
  let deeplink, deeplinks, new_link, parent, target;
  deeplink = 'link';
  deeplinks = document.querySelectorAll(`.${deeplink}`);
  if(deeplinks) {
    document.addEventListener('click', function(event)
    {
      target = event.target;
      parent = target.parentNode;
      if (target && target.classList.contains(deeplink) || parent.classList.contains(deeplink)) {
        event.preventDefault();
        new_link = target.href != undefined ? target.href : target.parentNode.href;
        navigator.clipboard.writeText(new_link);
        target.href != undefined ?  copyFeedback(target) : copyFeedback(target.parentNode);
      }
    });
  }
}

// Tables made responsive with CSS: table { overflow-x: auto; }

function backToTop(){
  const toTop = document.querySelector("#toTop");
  if (toTop) {
    window.addEventListener("scroll", () => {
      toTop.classList.toggle('active', window.scrollY >= 200);
    }, { passive: true });
  }
}

// Lazy loading handled by native HTML loading="lazy" attribute

function loadActions() {
  copyHeadingLink();
  backToTop();
}

// Critical path - initialize immediately
initializeNavigation();
// Use requestIdleCallback for non-critical tasks
if ('requestIdleCallback' in window) {
  requestIdleCallback(() => {
    loadActions();
  }, { timeout: 1000 });
} else {
  // Fallback for browsers without requestIdleCallback
  setTimeout(() => {
    loadActions();
  }, 0);
}
