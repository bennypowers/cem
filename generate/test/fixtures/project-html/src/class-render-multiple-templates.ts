import { LitElement, html } from 'lit-element';
import { customElement } from 'lit-element/decorators.js';

@customElement('class-render-multiple-templates')
class ClassRenderMultipleTemplates extends LitElement {
  render() {
    return [
      html`
        <!-- First Slot -->
        <slot name="first"></slot>
        <!-- Part One -->
        <div part="one"></div>
      `,
      html`
        <!-- Second Slot -->
        <slot name="second"></slot>
        <!-- Part Two -->
        <div part="two"></div>
      `
    ];
  }
}

@customElement('class-render-composed-templates')
class ClassRenderComposedTemplates extends LitElement {
  render() {
    const composedSlot = html`
      <!-- Composed Slot -->
      <slot name="composed"></slot>
    `;
    return html`
      ${composedSlot}
      <!-- Main Slot -->
      <slot name="main"></slot>
    `;
  }
}
