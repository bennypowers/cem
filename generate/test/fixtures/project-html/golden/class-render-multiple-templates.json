{
  "schemaVersion": "2.1.0",
  "modules": [
    {
      "kind": "javascript-module",
      "path": "src/class-render-multiple-templates.js",
      "declarations": [
        {
          "name": "ClassRenderMultipleTemplates",
          "superclass": {
            "name": "LitElement",
            "package": "lit"
          },
          "kind": "class",
          "tagName": "class-render-multiple-templates",
          "slots": [
            {
              "name": "first",
              "description": "First Slot"
            },
            {
              "name": "second",
              "description": "Second Slot"
            }
          ],
          "cssParts": [
            {
              "name": "one",
              "description": "Part One"
            },
            {
              "name": "two",
              "description": "Part Two"
            }
          ],
          "customElement": true
        },
        {
          "name": "ClassRenderComposedTemplates",
          "superclass": {
            "name": "LitElement",
            "package": "lit"
          },
          "kind": "class",
          "tagName": "class-render-composed-templates",
          "slots": [
            {
              "name": "composed",
              "description": "Composed Slot"
            },
            {
              "name": "main",
              "description": "Main Slot"
            }
          ],
          "customElement": true
        }
      ],
      "exports": [
        {
          "kind": "custom-element-definition",
          "name": "class-render-multiple-templates",
          "declaration": {
            "name": "ClassRenderMultipleTemplates",
            "module": "src/class-render-multiple-templates.js"
          }
        },
        {
          "kind": "custom-element-definition",
          "name": "class-render-composed-templates",
          "declaration": {
            "name": "ClassRenderComposedTemplates",
            "module": "src/class-render-multiple-templates.js"
          }
        }
      ]
    }
  ]
}
